File: Java/406.Queue_Reconstruction_by_Height.java
class Solution {
    public int[][] reconstructQueue(int[][] people) {
        int[][] sortedPeople = people;
        Arrays.sort(sortedPeople, new Comparator<int[]>() {
            @Override
            public int compare(int[] a, int[] b) {
                return a[0] != b[0] ? b[0] - a[0] : a[1] - b[1];
            }
        });
        List<int[]> res = new LinkedList<>();
        for (int[] person : sortedPeople) {
            res.add(person[1], person);
        }
        return res.toArray(new int[people.length][]);
    }
}

File: Python3/406.Queue_Reconstruction_by_Height.py
class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        sorted_heights = sorted(people)
        res = [None for _ in range(len(people))]
        for person in sorted_heights:
            ahead_pos = person[1]
            count = 0
            for i in range(len(res)):
                if (count == ahead_pos and res[i] == None):
                    res[i] = person
                    break
                if (res[i] == None or res[i][0] >= person[0]):
                    count += 1
        return res

# simpler solution but same time complexity


class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        sorted_people = sorted(people, key=lambda x: (-x[0], x[1]))
        res = []
        for person in sorted_people:
            res.insert(person[1], person)
        return res
